test/refactoring/ConvertAsync/convert_pattern.swift

```diff
@@ -5,6 +5,305 @@ enum E : Error { case e }
+    }
+  }
+
+  // MIXED-TUPLE-RESULT:      do {
+  // MIXED-TUPLE-RESULT-NEXT:   let res = try await mixedTupleResult()
+  // MIXED-TUPLE-RESULT-NEXT:   let ((x, y), z) = res
+  // MIXED-TUPLE-RESULT-NEXT:   let ((x1, _), z1) = res
+  // MIXED-TUPLE-RESULT-NEXT:   print("a", x, y, z)
+  // MIXED-TUPLE-RESULT-NEXT:   print("b", x, z)
+  // MIXED-TUPLE-RESULT-NEXT: } catch let err {
+  // MIXED-TUPLE-RESULT-NEXT:   print("oh no")
+  // MIXED-TUPLE-RESULT-NEXT:   print("oh no again")
+  // MIXED-TUPLE-RESULT-NEXT: }
+
+  // RUN: %refactor-check-compiles -convert-call-to-async-alternative -dump-text -source-filename %s -pos=%(line+1):3 | %FileCheck -check-prefix=MIXED-TUPLE-RESULT2 %s
+  mixedTupleResult { res in
+    switch res {
+    case .success(let ((x, _), z)):
+      print(x, z)
+    case .failure(let err):
+      print("oh no \(err)")
+    }
+  }
+
+  // MIXED-TUPLE-RESULT2:      do {
+  // MIXED-TUPLE-RESULT2-NEXT:   let ((x, _), z) = try await mixedTupleResult()
+  // MIXED-TUPLE-RESULT2-NEXT:   print(x, z)
+  // MIXED-TUPLE-RESULT2-NEXT: } catch let err {
+  // MIXED-TUPLE-RESULT2-NEXT:   print("oh no \(err)")
+  // MIXED-TUPLE-RESULT2-NEXT: }
+
+  // RUN: %refactor-check-compiles -convert-call-to-async-alternative -dump-text -source-filename %s -pos=%(line+1):3 | %FileCheck -check-prefix=MIXED-TUPLE-RESULT3 %s
+  mixedTupleResult { res in
+    if let ((_, y), z) = try? res.get() {
+      print(y, z)
+    } else {
+      print("boo")
+    }
+  }
+
+  // MIXED-TUPLE-RESULT3:      do {
+  // MIXED-TUPLE-RESULT3-NEXT:   let ((_, y), z) = try await mixedTupleResult()
+  // MIXED-TUPLE-RESULT3-NEXT:   print(y, z)
+  // MIXED-TUPLE-RESULT3-NEXT: } catch {
+  // MIXED-TUPLE-RESULT3-NEXT:   print("boo")
+  // MIXED-TUPLE-RESULT3-NEXT: }
+
+  // RUN: %refactor-check-compiles -convert-call-to-async-alternative -dump-text -source-filename %s -pos=%(line+1):3 | %FileCheck -check-prefix=MULTIPLE-TUPLE-PARAM %s
+  multipleTupleParam { strs, ints, err in
+    guard let (str1, str2) = strs, let (int1, int2) = ints else {
+      print("ohno")
+      return
+    }
+    print(str1, str2, int1, int2)
+  }
+  // MULTIPLE-TUPLE-PARAM:      do {
+  // MULTIPLE-TUPLE-PARAM-NEXT:   let ((str1, str2), (int1, int2)) = try await multipleTupleParam()
+  // MULTIPLE-TUPLE-PARAM-NEXT:   print(str1, str2, int1, int2)
+  // MULTIPLE-TUPLE-PARAM-NEXT: } catch let err {
+  // MULTIPLE-TUPLE-PARAM-NEXT:   print("ohno")
+  // MULTIPLE-TUPLE-PARAM-NEXT: }
+
+  // RUN: %refactor-check-compiles -convert-call-to-async-alternative -dump-text -source-filename %s -pos=%(line+1):3 | %FileCheck -check-prefix=MULTIPLE-TUPLE-PARAM2 %s
+  multipleTupleParam { strs, ints, err in
+    guard let (str1, str2) = strs, var (int1, int2) = ints else {
+      print("ohno")
+      return
+    }
+    print(str1, str2, int1, int2)
+  }
+  // MULTIPLE-TUPLE-PARAM2:      do {
+  // MULTIPLE-TUPLE-PARAM2-NEXT:   var ((str1, str2), (int1, int2)) = try await multipleTupleParam()
```

Comment:
What are your thoughts on keeping `strs`/`ints` and then binding `str*` and `int*` separately when there is mixed var/let?

Reply:
We could do that, though I wonder if it makes sense to apply more generally, e.g if its `var (x, y) = ...` but only `x` is a `var` in the closure body, should we generate the following?

```
let (x, y) = ...
var x = x
```

Reply:
Hmmm... that does seem a little weird 😆. Let's just leave as is I think.

Reply:
Okay, so it looked weird because that would be a redecl 🤦.

We get a warning if a variable is mutable and not mutated, which is a bit of a pain with the case patterns, ie. if we make `case (let x, var y) = res` into `var (x, y) = res` there would then be a warning for `x`. Probably okay as an initial implementation?

Reply:
Yeah it would actually need to something more like this for us to handle it:

```
let (x1, y) = ...
var x = x1
```

Given this is a more general thing, I think I'll leave it for now, but I'll file a radar for us to handle it.

Reply:
And turns out it isn't actually a warning, I just had other warnings re. no reads/writes. So... all good 👍